[RISCV] Add tune info for postra scheduling direction

wangpc-pp opened the pull request:
The results differ on different platforms so it is really hard to
determine a common default value.
    
Tune info for postra scheduling direction is added and CPUs can
set their own preferable postra scheduling direction.

preames reviewed (commented):
At least for me, the diff here is too large to meaningfully review.  I think we need an alternate form of justification that this is profitable.  Can you present either some performance numbers or at minimum some kind of statistic that demonstrates value?

For context, I'm not particularly skeptical of the patch - it seems to make sense - the diffs are just way too big to be meaningfully skimmed.

michaelmaitland commented:
How are we evaluating this change? On spills? On impact to dynamic IC? On impact to runtime on real hardware?

wangpc-pp commented:
> How are we evaluating this change? On spills? On impact to dynamic IC? On impact to runtime on real hardware?

IIUC, PostRA scheduling basically won't impact on spills and dynamic instruction count. I will show some cycles/IPC based on GEM5.
I do agree this should be tuned by CPUs, I will make it a tuning feature later but we still need a default setting (`topdown`, `buttomup` or `bidirectional`, this can be debated).

wangpc-pp committed:
[RISCV] Add tune info for postra scheduling direction

The results differ on different platforms so it is really hard to
determine a common default value.

Tune info for postra scheduling direction is added and CPUs can
set their own preferable postra scheduling direction.

wangpc-pp commented:
I got some results on different platforms.

1. Coremark with 60000 iterations on GEM5: `-O3 -march=rv64gc` (use this configuration so there are more instructions to schedule and GEM5 has a better support)

  * RiscvO3CPU

|        | Baseline    | TopDown     | BottomUp    | Bidirectional |
|--------|-------------|-------------|-------------|---------------|
| Cycles | 33140248616 | 33031782179 | 33173323700 | 33063777692   |
| IPC    | 1.381871    | 1.386409    | 1.380822    |   1.383095   |

  * RiscvAtomicSimpleCPU

|        | Baseline    | TopDown     | BottomUp    | Bidirectional |
|--------|-------------|-------------|-------------|---------------|
| Cycles | 30563561357 | 30935303916 | 30898754397 | 30563777692   |
| IPC    | 0.879774    | 0.869409    | 0.870437    | 0.879768      |

2. Coremark on Spacemit-X60 (`-O3 -mcpu=spacemit-x60`):

|        | Baseline    | TopDown     | BottomUp    | Bidirectional |
|--------|-------------|-------------|-------------|---------------|
| Iterations/Sec| 4763.71 | 4993.34 | 4730.50 | 4853.86 |

The results differ on different platforms so it is really hard to determine a default value. I will make it a target feature and leave the default value to be `Bidirectional` (this can be debated).

mshockwave commented:
> I got some results on different platforms.
> 
> 1. Coremark with 60000 iterations on GEM5: `-O3 -march=rv64gc` (use this configuration so there are more instructions to schedule and GEM5 has a better support)
> 
> * RiscvO3CPU
> 
> Baseline	TopDown	BottomUp	Bidirectional
> Cycles	33140248616	33031782179	33173323700	33063777692
> IPC	1.381871	1.386409	1.380822	1.383095
> * RiscvAtomicSimpleCPU
> 
> Baseline	TopDown	BottomUp	Bidirectional
> Cycles	30563561357	30935303916	30898754397	30563777692
> IPC	0.879774	0.869409	0.870437	0.879768
> 2. Coremark on Spacemit-X60 (`-O3 -mcpu=spacemit-x60`):
> 
> Baseline	TopDown	BottomUp	Bidirectional
> Iterations/Sec	4763.71	4993.34	4730.50	4853.86
> The results differ on different platforms so it is really hard to determine a default value. I will make it a target feature and leave the default value to be `Bidirectional` (this can be debated).

I have no objection making it into a tuning feature.

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1438,6 +1438,20 @@ def TunePreferWInst
     : SubtargetFeature<"prefer-w-inst", "PreferWInst", "true",
                        "Prefer instructions with W suffix">;
 
+let SetMaxValue = false in {
```
Can we put this in TuneInfo instead? Seems wasteful to allocate 3 subtarget bits for something that already has command line options in the scheduler. We just need to be able to map CPU to a value just like cache size or prefetch distance right?

wangpc-pp replied:
Good idea! I forgot about TuneInfo. Thanks!

topperc commented:
Let me make sure I understand. BottomUp and bi-directional RA scheduling were added last year by @michaelmaitland 9106b58ce4e8dada167eec50178a9e154342e4ba. I don't think we ended up enabling inside SiFive. Have any other targets adopted it yet?

wangpc-pp commented:
> Let me make sure I understand. BottomUp and bi-directional RA scheduling were added last year by @michaelmaitland [9106b58](https://github.com/llvm/llvm-project/commit/9106b58ce4e8dada167eec50178a9e154342e4ba). I don't think we ended up enabling inside SiFive. Have any other targets adopted it yet?

Not yet. And I just found some issues when enabling postra bidirectional scheduling: #116592 and #116584.
Or we can set it to topdown by default, all current data show that it has the best result.

topperc commented:
> Member

Let's go with top down by default.

wangpc-pp commented:
> > Member
> 
> Let's go with top down by default since that's what other target use.

Done.

michaelmaitland reviewed (approved):
LGTM

topperc reviewed (approved):
LGTM
